students/JerryH/Lesson04/mailroom2.py

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python3
+
+# List of sample donors and donations
+donors = {
+    "Bill Gates": [234.22, 45645.24, 43953.09, 98823],
+    'Jeff Bezo': [4564.23],
+    'Mike Dell': [299.09, 26273.67],
+    'Harry Potter': [8234.09, 83948.04, 7834.23],
+    'Ben Williams': [29283.00, 1334.34],
+    'Guy James': [93.00, 34.34]
+    }
+
+options = {
+    'THANK_YOU_OPT': 1,
+    'REPORT_OPT': 2,
+    'SEND_LETTER_OPT': 3,
+    'QUIT_OPT': 4
+    }
+
+menu = {
+    'op1': "Send a Thank You",
+    'op2': "Create a Report",
+    'op3': "Send Letters To Everyone",
+    'op4': "Quit"
+    }
+
+
+def prompt():
+    return int(input("Please choose the following options:\n1) {op1}.\n2) {op2}.\n3) {op3}.\n4) {op4}\n".format(**menu)))
+
+
+def donor_name_prompt():
+    return input("Send a Thank You - Please enter a full name or type \"list\""
+        "to list the current donors:\n")
+
+
+def get_all_donor_names():
+    all_names = []
+    for d in donors.keys():
+        all_names.append(d)
+    return all_names
+
+
+def list_all_donor_names():
+    for each_name in sorted(get_all_donor_names()):
+        print("{}".format(each_name))
+
+
+def main():
+    op = 0
+    while op != options['QUIT_OPT']:
+        op = prompt()
+
+        if op == options['THANK_YOU_OPT']:  # Append or list donors
+            donor_name = None
+            while not donor_name:
+                donor_name = donor_name_prompt()
+                if donor_name.lower() == "list":
+                    donor_name = None
+                    list_all_donor_names()
+            # print(donor_name)
+            donation = float(input("Please enter the donation amount:\n"))
+            # If the donnor doesn't exist in the donor dictionary - add his info
+            if donor_name not in get_all_donor_names():
+                # print(donors)
+                donors[donor_name] = [donation]
+            # else append the donation into the existing donors
+            else:
+                donors[donor_name].append(donation)
```

Comment:
I'm getting a little fancy here 😃 

there is an "extension" of a `dict` object called `defaultdict` that I'm really fond of, it makes some things much easier and cleaner.
The way it works is at the time you create your dict you specify default value, so in your case instead of having to check if donor name exists so you can safely use `append` on the existing donor, you just define default value as list and go ahead and use `append` regardless.

you can read the doc here:
https://docs.python.org/2/library/collections.html#collections.defaultdict

and see my example below:

```
>>> donors = defaultdict(list, {
...     "Bill Gates": [234.22, 45645.24, 43953.09, 98823],
...     'Jeff Bezo': [4564.23],
...     'Mike Dell': [299.09, 26273.67],
...     'Harry Potter': [8234.09, 83948.04, 7834.23],
...     'Ben Williams': [29283.00, 1334.34],
...     'Guy James': [93.00, 34.34]
...     })
>>>
>>> donors["James Bond"].append(15000)
>>> donors
defaultdict(<class 'list'>, {'Bill Gates': [234.22, 45645.24, 43953.09, 98823], 'Jeff Bezo': [4564.23], 'Mike Dell': [299.09, 26273.67], 'Harry Potter': [8234.09, 83948.04, 7834.23], 'Ben Williams': [29283.0, 1334.34], 'Guy James': [93.0, 34.34], 'James Bond': [15000]})
```